### PR TITLE
tests: interrupt: add test cases for testing direct interrupt of arch x86 and posix

### PR DESCRIPTION
**Pull request by enjiamai**
Add test case of direct interrupt for arch x86 and posix.  The test case is designed as follows:

1. Register two direct interrupt at build time.
2. Trigger interrupt and check if ISR handler has executed or not.
3. Also check irq_enable and irq_disable works.

Why we add an extra compiler option "-mgeneral-regs-only" to make it works in arch x86. Because there are some existing x87 instructions inside interrupt context. If we do not add GCC compiler option "-mgeneral-regs-only" as configuration when we use direct interrupt in arch x86, otherwise, a compilation error will happen: "sorry, unimplemented: 80387 instructions aren't allowed in an interrupt service routine".

### CODE REVIEW

**Comment by andyross** [comment]
Sorry, where is the FPU code that this is tripping over?   If that's real, we need to fix that much more urgently than we need to merge this test.

**Comment by enjiamai** [comment]
> Sorry, where is the FPU code that this is tripping over? If that's real, we need to fix that much more urgently than we need to merge this test.

Hi @andyross  the error message is like this when we did not add  -mgeneral-regs-only
/home/mm/zephyrproject/zephyr/tests/kernel/interrupt/src/direct_isr.c:49:1: sorry, unimplemented: 80387 instructions aren't allowed in an interrupt service routine 
   49 | ISR_DIRECT_DECLARE(direct_isr1)

And I remembered during my trial, I ever saw some message related to SSE,  seems like in lib/os/cbprintf.c  
        va_start(ap, format);
        rc = cbvprintf(out, ctx, format, ap);
        va_end(ap);

Sorry, I did not remember how I make this message happened, let me try to reproduce it.

**Comment by LeiW000** [comment]
Please address the "Compliance" failure.

**Comment by enjiamai** [comment]
> Please address the "Compliance" failure.

Ok, @LeiW000 got it. Got it fixed. Thank you!

**Comment by enjiamai** [comment]
Hi,  @andyross 

I really did not find any FPU code here. I also tried to define an isr handler myself like this:
```
__attribute__ ((interrupt)) void direct_isr1(void *stack_frame)
{
        //ARG_UNUSED(stack_frame);

        //ISR_DIRECT_HEADER();

        //direct_int_executed[0]++;

        //ISR_DIRECT_FOOTER(0);
}
```

Even I comment all all code inside this function, it still shows following message:

```
/home/mm/zephyrproject/zephyr/tests/kernel/interrupt/src/direct_isr.c:48:1: sorry, unimplemented: 80387 instructions aren't allowed in an interrupt service routine
   48 | {
      | ^
```

I found two ways can avoid this compilation error:
1. Add the "-mgeneral-regs-only" gcc option for x86.
2. Remove the __attribute__ ((interrupt)) in "include/arch/x86/ia32/arch.h" line:316 
(This means in the exit of this function will use RET instead of IRET. I am not sure doing this is acceptable)

I was stuck here for a while, could you please give me some advice, thanks!

**Comment by andyross** [comment]
OK, I think I get it.  We have code generation set to enable floating point, so this function (or any that it calls) can legally build with float or double variables and emit 8087 instructions.  But GCC assumes that FPU exceptions can't be handled in an interrupt (this isn't actually true in the general case, you CAN have a scheme that allows this in theory, but it's not common OS design).

The GCC warning is actually correct as far as it goes.  We don't have any protection in our x86 arch layer to prevent kernel code in an interrupt from calling into something that tries the hit the FPU.  We just assume that it won't because we didn't write any such code.  But the compiler can't know that.

Dig around, there is very likely a code generation attribute that you can apply to this particular function that will disable FPU generation on this one particular function and thus suppress the warning.

Alternatively, we have to fix this at the level of the direct interrupt feature.   We could hide it behind a "CONFIG_X86_ENABLE_DIRECT_IRQ" or something, and when that is true disable FPU registers everywhere.  Which seems... not awful I guess.

Alternatively alternatively, we could drop the direct interrupt feature from the architecture entirely.  It's obviously not compatible with the MMU or userspace, I don't think we ever made it work on x86_64, it doesn't play well with the scheduler (we don't document it, but direct interrupts always return to the interrupted thread and thus can't wake a high priority thread up or suspend _current) and frankly super-low-latency interrupts aren't a x86 design goal on any current platforms.  @dcpleung might have input regarding 32 bit x86 needs.

**Comment by enjiamai** [comment]
Wow, it was a kind of epiphany, thank you so much! @andyross 
For x86_64 in our design, we only have dynamic isr in place.
For x86, our codebase has all these three kinds of interrupts implemented, but so far I did not see the existing test cases for the IRQ_DIRCT_CONNECT (or IRQ_CONNECT), that's why I am trying to add them here.

But your words let me understand there are indeed other design considerations for this. 

So it seems like we might at least have two choices here:  
1. Add a dedicated CONFIG_X86_ENABLE_DIRECT_IRQ for enabling this and without using FPU to ensure system security. Although it might have some problems or potential security risks. 
2. Do not use direct isr for x86. I was thinking if this is the option, from QA's perspective, we might need to remove the existing code of this, otherwise, we cannot achieve the goal of the code coverage. 

And hope for some input about x86 from @dcpleung , thanks!

**Comment by dcpleung** [comment]
No one uses direct interrupt on x86, AFAIK.

**Review by LeiW000 — APPROVED** [review]
Approve it. The code of DIRECT_IRQ should be kept.

File: tests/kernel/interrupt/src/direct_isr.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+#endif
+
+volatile uint32_t direct_int_executed[2];
+
+ISR_DIRECT_DECLARE(direct_isr1)
+{
+	direct_int_executed[0]++;
+
+	return 0;
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/direct_isr.c:51]: Should be there `return 1` too like in the macro below?

File: tests/kernel/interrupt/src/direct_isr.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	irq_disable(TEST_DIRECT_IRQ_LINE_2);
+
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_1);
+	trigger_irq(TRIGGER_DIRECT_IRQ_LINE_2);
+
+	/*
+	 * irq_enable()/irq_disable() does not work here,
+	 * see #33901.
+	 */
+#if !defined(CONFIG_X86)
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/direct_isr.c:128]: I'm not sure if we can allow to merge the test with such workaround. I had a bit of look at issue #33901. Seems like no further discussions about it.
@enjiamai you have to ask  @andyross about further solution of the #33901

File: tests/kernel/interrupt/src/direct_isr.c (hunk 1 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/direct_isr.c:51]: Hi @maksimmasalski , it should return zero here. There return value decides if a swap should happen after direct ISR. You can check "void arch_isr_direct_footer(int swap)" for more details, thanks!

File: tests/kernel/interrupt/src/direct_isr.c (hunk 2 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/direct_isr.c:128]: Yes, @maksimmasalski, let me ask him tonight.  And you talk about a point. I really want to make sure that can the testcases be merged even the CI does not pass due to a bug? Thanks so much for your comments.

Comment (maksimmasalski) [reply tests/kernel/interrupt/src/direct_isr.c:128]: Yes, ask him please. Test case looks decent, but that workaround is something that necessary to solve.

**Review by andyross — APPROVED** [review]
I guess I lost my argument that we should just drop this feature on x86.  But the test looks clean and I like the way the compiler argument is isolated.

**Review by jenmwms — APPROVED** [review]

**Review by maksimmasalski — APPROVED** [review]

**Pull request merged**